### DISCUSSION TITLE
Using AppThemeResource with ToolbarItem IconImageSource Invalid Resource Type

### DISCUSSION THREAD

**Discussion by Orgbrat** [CommunityToolkit/Maui#1940] [Q&A]
So I was doing this in .Net 7 with NO issues I thought but since moving to .Net 8 I am receiving and error on  setting the ToolbarItem IconImageSource.  Has anyone tried this and it works or am I doing something wrong? Or is there a better way to switch images by theme now?

Error : Invalid resource type: expected type is ImageSource, actual type is 'AppThemeObject'

`        <toolkit:AppThemeObject
            x:Key="MyAddImageSource"
            Dark="symboladdwhite.png"
            Light="symboladd.png" />`

`        <ToolbarItem
            Command="{Binding AddAppointmentCommand}"
            IconImageSource="{toolkit:AppThemeResource MyAddImageSource}"
            Text="New" />`

Thanks

**Comment by bijington** [comment]
I think I remember something around changes to type expectations like this. I think you will need to subclass `AppThemeObject` with something like: `public class AppThemeImageSource : AppThemeObject<InageSource>`

**Reply by Orgbrat** [reply]
@bijington  Thank you Shaun for the quick reply. I will try this and see what happens.

**Reply by Orgbrat** [reply]
@bijington Oh no, spoke too soon...    

I added the subclass and changed the page resource to the subclass like 
     `<controls:AppThemeImageSource
            x:Key="MyAddImageSource"
            Dark="symboladdwhite.png"
            Light="symboladd.png" />`

But now the error is :  Error : Invalid resource type: expected type is ImageSource, actual type is 'AppThemeImageSource'
How do I get the AppThemeResource  `IconImageSource="{toolkit:AppThemeResource MyAddImageSource}"` to be ok with the sub-classed class?  Has to be a way to do this as it worked great in .Net 7.  I have scoured the internet and the CommunityToolkit.Maui docs and have found nothing to help or assist.

**Reply by bijington** [reply]
Hmm looking inside the `AppThemeResource` code https://github.com/CommunityToolkit/Maui/blob/2402be17ee7f0c3cf9f3e42c75316905c998fd1f/src/CommunityToolkit.Maui/Extensions/AppThemeResourceExtension.shared.cs#L32 I can't see how we support custom implementations. @jfversluis do you know how we can potentially help @Orgbrat to solve this?

**Reply by bijington** [reply]
That said I also can't see the error that you are receiving in the toolkits codebase

**Reply by Orgbrat** [reply]
@bijington I looked at AppThemeObject.shared.cs and the AppThemeResourceExtension.shared.cs code and I can't see how to make this work either. I hope @jfversluis has some idea...

**Reply by bijington** [reply]
I'm not suggesting this is the solution but a further step you could try is to copy the AppThemeResourceExtension source code and see if you can add an extra clause into the ProvideValue method supporting your type. If that works we should look to open this up somehow to make it work

**Reply by jfversluis** [reply]
Interesting, does it work with a regular Image? Is it specific to the ToolbarItem? I wonder if something changed on the .NET MAUI side for this?

**Reply by Orgbrat** [reply]
@jfversluis @bijington  Do not know if it works with a regular Image Source.  But I have been using this with the ToolbarItem IconImageSource property all over my app and it did work with no issues in .Net 7. According to the .NET docs this is a standard Maui ImageSource. Something must have changed in .NET 8. As I see no way to do what @bijington  suggests and it work without an error.

**Reply by jfversluis** [reply]
Tried to look into this, am I missing something here? https://github.com/jfversluis/AppThemeObjectTest

This seems to compile and from what I can determine with the poor image I have in there seems to be working?

**Reply by Orgbrat** [reply]
@jfversluis  Well it compiles for me too but as you can see it has the red squiggle line and the image does not appear at runtime.
And if you hover over red squiggle line it states :  Error : Invalid resource type: expected type is ImageSource, actual type is 'AppThemeObject'. I have so far not determined what it wants from me in the XAML code.

![image](https://github.com/CommunityToolkit/Maui/assets/774141/367aa692-20a7-4cc8-82ef-3946dacebfb5)

Steve

**Reply by bijington** [reply]
@Orgbrat are you testing this on iOS/macOS? If so I see the same behaviour:


https://github.com/CommunityToolkit/Maui/assets/17139988/a44ad8fd-53f7-46ab-9b6b-7c38b06ce3ee

**Reply by bijington** [reply]
You can see that the button ImageSource property updates as expected but the ToolbarItem IconImageSource does not

**Reply by bijington** [reply]
But then if I remove the icon and just rely on the text I see the same behaviour which makes me believe this is a bug in the ToolbarItem and not the AppThemeResource

https://github.com/CommunityToolkit/Maui/assets/17139988/01ec6eaa-6941-4daa-8196-98389bd89ac6

**Reply by Orgbrat** [reply]
@bijington Actually I am testing on Windows  Also if this is a ToolbarItem bug, I believe it was a bug introduced in .NET 8.0 as I am almost certain it worked in .NET 7.0 just fine. But I could be wrong...  :)

**Comment by Orgbrat** [comment]
So this code still has issues after migrating to latest code bits  MCT 10.0.0 and .Net 9.0. It shows same red squiggly with same error.  Does anyone have any idea how to resolve this error. Or how to use different images for light and dark at runtime?  Is this possible and if so what am I doing wrong...

**Comment by Orgbrat** [comment]
Ok, I am just going to perform the dynamic change in the page code-behind. Just is not going to work in the page XAML.
Thanks for all your help.
